src/vaults/VaultInternal.ts

```diff
@@ -536,6 +537,86 @@ class VaultInternal {
     });
   }
 
+  /**
+   * Acquire a read-only lock on this vault
+   */
+  @ready(new vaultsErrors.ErrorVaultNotRunning())
+  public acquireRead(): ResourceAcquire<FileSystemReadable> {
+    return async () => {
+      const acquire = this.lock.read();
+      const [release] = await acquire();
+      return [
+        async (e?: Error) => {
+          await release(e);
+        },
+        this.efsVault,
+      ];
+    };
+  }
+
+  /**
+   * Acquire a read-write lock on this vault
+   */
+  @ready(new vaultsErrors.ErrorVaultNotRunning())
+  public acquireWrite(
+    tran?: DBTransaction,
+  ): ResourceAcquire<FileSystemWritable> {
+    return async () => {
+      let releaseTran: ResourceRelease | undefined = undefined;
+      const acquire = this.lock.write();
+      const [release] = await acquire();
+      if (tran == null) {
+        const acquireTran = this.db.transaction();
+        [releaseTran, tran] = await acquireTran();
+      }
+      // The returned transaction can be undefined, too. We won't handle those
+      // cases.
+      if (tran == null) utils.never('Acquired transactions cannot be null');
+      await tran.lock(
+        [...this.vaultMetadataDbPath, VaultInternal.dirtyKey].join(''),
+      );
+      if (
+        (await tran.get([
+          ...this.vaultMetadataDbPath,
+          VaultInternal.remoteKey,
+        ])) != null
+      ) {
+        // Mirrored vaults are immutable
+        throw new vaultsErrors.ErrorVaultRemoteDefined();
+      }
+      await tran.put(
+        [...this.vaultMetadataDbPath, VaultInternal.dirtyKey],
+        true,
+      );
+      return [
+        async (e?: Error) => {
+          if (e == null) {
+            try {
+              // After doing mutation we need to commit the new history
+              await this.createCommit();
+            } catch (e_) {
+              e = e_;
+            }
+            // This would happen if an error was caught inside the catch block
+            if (e != null) {
+              // Error implies dirty state
+              await this.cleanWorkingDirectory();
+            }
```

Comment:
Why not just put the `await this.cleanWorkingDirectory();`  inside the catch block then?